### PR TITLE
Senlin: Nodes Create

### PR DESCRIPTION
**Pull request by JackKuei**
Prior to a PR being reviewed, there needs to be a Github issue that the PR
addresses. Replace the brackets and text below with that issue number.

For #[823 [Add support for Senlin clustering]](https://github.com/gophercloud/gophercloud/issues/823)

Links to the line numbers/files in the OpenStack source code that support the
code in this PR:

[[nodes:create]](https://github.com/openstack/senlin/blob/e5bc16b3fea2f10dead365f983c5a937f23d72b0/senlin/api/openstack/v1/nodes.py#L61)

### CODE REVIEW

**Comment by JackKuei** [comment]
Dependent on #871 and #872

**Comment by JackKuei** [comment]
recheck

**Comment by jtopjian** [comment]
@JackKuei You don't need to recheck the OpenLab tests -- they aren't running Selin tests, so any failures you see are not part of your code. I'll let you know if/when OpenLab begins testing Senlin :)

**Comment by JackKuei** [comment]
Sounds good....just like to see green...:)

File: openstack/clustering/v1/nodes/requests.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	ToNodeCreateMap() (map[string]interface{}, error)
+}
+
+// CreateOpts params
+type CreateOpts struct {
+	Role      string                 `json:"role,omitempty"`
+	ProfileID string                 `json:"profile_id" required:"true"`
+	ClusterID string                 `json:"cluster_id,omitempty"`
+	Name      string                 `json:"name" required:"true"`
+	Metadata  map[string]interface{} `json:"metadata,omitempty"`
```
[the remark this answers is not in the document]
Comment (JackKuei) [reply openstack/clustering/v1/nodes/requests.go]: From the doc ```A collection of key-value pairs associated with an object.``` Added test cases to have different value types such as bool, string, float

File: openstack/clustering/v1/nodes/results.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+type Node struct {
+	ClusterID    string                 `json:"cluster_id"`
+	CreatedAt    time.Time              `json:"-"`
+	Data         DataType               `json:"data"`
+	Dependents   map[string]interface{} `json:"dependents"`
+	Domain       string                 `json:"domain"`
+	ID           string                 `json:"id"`
+	Index        int                    `json:"index"`
+	InitAt       time.Time              `json:"-"`
+	Metadata     map[string]interface{} `json:"metadata"`
```
[the remark this answers is not in the document]
Comment (JackKuei) [reply openstack/clustering/v1/nodes/results.go]: Done

File: openstack/clustering/v1/nodes/results.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+type DataType struct {
+	InternalPorts []InternalPort `json:"internal_ports"`
+	Placement     Placement      `json:"placement"`
+}
+
+// Node represents a node structure
+type Node struct {
+	ClusterID    string                 `json:"cluster_id"`
+	CreatedAt    time.Time              `json:"-"`
+	Data         DataType               `json:"data"`
```
[the remark this answers is not in the document]
Comment (JackKuei) [reply openstack/clustering/v1/nodes/results.go]: Maybe should use map[string]interface{}?

https://github.com/openstack/senlin/blob/36ec6ef7bfed388272344bb6b32dc7532638d5de/senlin/profiles/os/nova/server.py#L769
https://github.com/openstack/senlin/blob/36ec6ef7bfed388272344bb6b32dc7532638d5de/senlin/profiles/os/nova/server.py#L801

[the remark this answers is not in the document]
Comment (JackKuei) [reply openstack/clustering/v1/nodes/results.go]: https://github.com/openstack/senlin/blob/8ba34d5997195c2da19a9f278cc07f7941add0e0/senlin/profiles/os/nova/server.py#L717

**Comment by JackKuei** [comment]
Here are the relevant links where Data is defined

https://github.com/openstack/senlin/blob/8ba34d5997195c2da19a9f278cc07f7941add0e0/senlin/profiles/os/nova/server.py#L717
https://github.com/openstack/senlin/blob/36ec6ef7bfed388272344bb6b32dc7532638d5de/senlin/profiles/os/nova/server.py#L769
https://github.com/openstack/senlin/blob/36ec6ef7bfed388272344bb6b32dc7532638d5de/senlin/profiles/os/nova/server.py#L801
https://github.com/openstack/senlin/blob/8ba34d5997195c2da19a9f278cc07f7941add0e0/senlin/profiles/os/nova/server.py#L66
https://github.com/openstack/senlin/blob/8ba34d5997195c2da19a9f278cc07f7941add0e0/senlin/profiles/os/nova/server.py#L868

**Comment by animationzl** [comment]
recheck

**Comment by animationzl** [comment]
recheck

**Comment by jtopjian** [comment]
> Here are the relevant links where Data is defined

Thank you :)

It looks like `Data` doesn't have a strict schema, which is ok -- it's not the first we've run into this. But I want to make sure we're not going to run into any unmarshal errors.

For example, I see there's a container/docker profile:

https://github.com/openstack/senlin/blob/8ba34d5997195c2da19a9f278cc07f7941add0e0/senlin/profiles/container/docker.py

If I were to create a container node based off of that profile, would its data adhere to `Data`?

**Comment by JackKuei** [comment]
Good question. From that docker profile code, seems like it is still in experimental https://github.com/openstack/senlin/blob/8ba34d5997195c2da19a9f278cc07f7941add0e0/senlin/profiles/container/docker.py#L35 https://github.com/openstack/senlin/blob/8ba34d5997195c2da19a9f278cc07f7941add0e0/senlin/profiles/container/docker.py#L154
But let me post a question in the senlin forum.

**Comment by dkt26111** [comment]
@jtopjian @JackKuei: The `Data` field is not set for the docker profile as @JackKuei mentioned.  So there should be no unmarshal errors when using docker profiles.
Having said that, I think it might be better to just define `Data` as `map[string]interface{}` to avoid having to make gophercloud changes every time the senlin code changes to introduce additional data fields.

**Comment by JackKuei** [comment]
@jtopjian Ready for review

**Pull request closed without merging**